zproject/backends.py

```diff
@@ -898,6 +899,94 @@ def external_auth_method(cls: Type[ExternalAuthMethod]) -> Type[ExternalAuthMeth
     EXTERNAL_AUTH_METHODS.append(cls)
     return cls
 
+# We want to be able to store this data in redis, so it has to be easy to serialize.
+# That's why we avoid having fields that could pose a problem for that.
+ExternalAuthDataDict = TypedDict('ExternalAuthDataDict', {
+    'subdomain': str,
+    'full_name': str,
+    'email': str,
+    'is_signup': bool,
+    'redirect_to': str,
+    'mobile_flow_otp': Optional[str],
+    'desktop_flow_otp': Optional[str],
+    'multiuse_object_key': str,
+    'full_name_validated': bool,
+}, total=False)
+
+class ExternalAuthResult:
+    LOGIN_KEY_PREFIX = "login_key_"
+    LOGIN_KEY_FORMAT = LOGIN_KEY_PREFIX + "{token}"
+    LOGIN_KEY_EXPIRATION_SECONDS = 15
+    LOGIN_TOKEN_LENGTH = UserProfile.API_KEY_LENGTH
+
+    def __init__(self, *, user_profile: Optional[UserProfile]=None,
+                 data_dict: Optional[ExternalAuthDataDict]=None,
+                 login_token: Optional[str]=None,
+                 delete_stored_data: bool=True) -> None:
+        if data_dict is None:
+            data_dict = {}
+
+        if login_token is not None:
+            assert (not data_dict) and (user_profile is None), ("Passing in data_dict or user_profile " +
+                                                                "with login_token is disallowed.")
+            self.instantiate_with_token(login_token, delete_stored_data)
+        else:
+            self.data_dict = cast(ExternalAuthDataDict, {**data_dict})
+            self.user_profile = user_profile
+
+        if self.user_profile is not None:
+            # Ensure data inconsistent with the user_profile wasn't passed in inside the data_dict argument.
+            assert 'full_name' not in data_dict or data_dict['full_name'] == self.user_profile.full_name
+            assert 'email' not in data_dict or data_dict['email'] == self.user_profile.delivery_email
+            # Update these data_dict fields to ensure consistency with self.user_profile. This is mostly
+            # defensive code, but is useful in these scenarios:
+            # 1. user_profile argument was passed in, and no full_name or email_data in the data_dict arg.
+            # 2. We're instantiating from the login_token and the user has changed their full_name since
+            #    the data was stored under the token.
+            self.data_dict['full_name'] = self.user_profile.full_name
+            self.data_dict['email'] = self.user_profile.delivery_email
```

Comment:
I guess I don't understand all this work.  My understanding of the situation is that IF we have a UserProfile, there are two cases:
* The user is a mirror dummy, and we're doing account creation, in which case we shouldn't copy data from `self.user_profile` at all; the intent is to replace the name of the dummy user with the name the user picks in the registration flow.
* The user is a real account, and we're logging in, in which case I don't think we expect to access either of these fields later in the flow, and we could just delete them from the object rather than doing this validation.
 
Am I missing something here?

Reply:
We need the dict for storing the information in redis when authing through the `auth` subdomain. So we need to put the email (at least, and if we're doing that, then it makes sense to take care of the full_name too) in the data_dict. Of course, we could simply overwrite those fields with values from `user_profile` when storing the dict, but that's what the `_load_user_data_from_user_profile` approach was doing and it was ugly too (uglier in my opinion).

Reply:
I'd love to clean this up further - one idea would be to simply store the userprofile `id` when the user is supposed to get logged in rather, but such changes would be out of scope of this commit and make it unreadable I think.

Reply:
Yeah, that's probably the right answer.  I'm OK with doing that cleanup in a follow-up commit.  We could even make two subclasses of ExternalAuthResult, one for `ExternalAuthResultLogin` and another `ExternalAuthResultSignup`, or something, if it'd be helpful to type things that way?  Not sure if that's a good idea, but it might be.